mypy/main.py

```diff
@@ -205,6 +207,72 @@ def invert_flag_name(flag: str) -> str:
+        return ['python{}'.format(v)]
+
+
+def _python_version_from_executable(python_executable: str) -> Tuple[int, int]:
+    try:
+        check = subprocess.check_output([python_executable, '-c',
+                                         'import sys; print(repr(sys.version_info[:2]))'],
+                                        stderr=subprocess.STDOUT).decode()
+        return ast.literal_eval(check)
+    except (subprocess.CalledProcessError, FileNotFoundError):
+        raise PythonExecutableInferenceError(
+            'Error: invalid Python executable {}'.format(python_executable))
+
+
+def _python_executable_from_version(python_version: Tuple[int, int]) -> str:
+    if sys.version_info[:2] == python_version:
+        return sys.executable
+    str_ver = '.'.join(map(str, python_version))
+    print(str_ver)
+    try:
+        sys_exe = subprocess.check_output(python_executable_prefix(str_ver) +
+                                          ['-c', 'import sys; print(sys.executable)'],
+                                          stderr=subprocess.STDOUT).decode().strip()
+        return sys_exe
+    except (subprocess.CalledProcessError, FileNotFoundError):
+        raise PythonExecutableInferenceError(
+            'Error: failed to find a Python executable matching version {},'
+            ' perhaps try --python-executable, or --no-site-packages?'.format(python_version))
+
+
+def infer_python_version_and_executable(options: Options,
+                                        special_opts: argparse.Namespace
+                                        ) -> Options:
+    # Infer Python version and/or executable if one is not given
+    if special_opts.python_executable is not None and special_opts.python_version is not None:
+        py_exe_ver = _python_version_from_executable(special_opts.python_executable)
+        if py_exe_ver != special_opts.python_version:
+            raise PythonExecutableInferenceError(
+                'Python version {} did not match executable {}, got version {}.'.format(
+                    special_opts.python_version, special_opts.python_executable, py_exe_ver
+                ))
+        else:
+            options.python_version = special_opts.python_version
+            options.python_executable = special_opts.python_executable
+    elif special_opts.python_executable is None and special_opts.python_version is not None:
+        options.python_version = special_opts.python_version
+        py_exe = None
+        if not special_opts.no_executable:
+            py_exe = _python_executable_from_version(special_opts.python_version)
+        options.python_executable = py_exe
+    elif special_opts.python_version is None and special_opts.python_executable is not None:
+        options.python_version = _python_version_from_executable(
+            special_opts.python_executable)
+        options.python_executable = special_opts.python_executable
+    return options
```

Comment:
Perhaps misleading that this modifies in place _and_ returns the modified value

Reply:
Are you suggesting that the comment at the beginning might be misleading? (That could be, I'm just not sure what you mean by this comment).

Reply:
I'd expect either
```
options.python_version, options.python_executable = infer_python_version_and_executable(special_opts)
```
or
```
infer_python_version_and_executable(options, special_opts)
```

But not
```
options = infer_python_version_and_executable(options, special_opts)
```

(think how `list.sort()`, `dict.update()`, etc, all modify in place and do not return the new result)